packages/wrangler/src/api/startDevWorker/RemoteRuntimeController.ts

```diff
@@ -1,30 +1,303 @@
+							assetDirectory: "",
+							excludePatterns: config.site.exclude ?? [],
+							includePatterns: config.site.include ?? [],
+						}
+					: undefined,
+				format: bundle.entry.format,
+				// TODO: Remove this passthrough
+				bindings: config._bindings ? config._bindings : bindings,
+				compatibilityDate: config.compatibilityDate,
+				compatibilityFlags: config.compatibilityFlags,
+				usageModel: config.usageModel,
+				routes,
+			});
+
+			// If we received a new `bundleComplete` event before we were able to
+			// dispatch a `reloadComplete` for this bundle, ignore this bundle.
+			// If `token` is undefined, we've surfaced a relevant error to the user above, so ignore this bundle
+			if (id !== this.#currentBundleId || !token) {
+				return;
+			}
+
+			const accessToken = await getAccessToken(token.host);
+
+			this.emitReloadCompleteEvent({
+				type: "reloadComplete",
+				bundle,
+				config,
+				proxyData: {
+					userWorkerUrl: {
+						protocol: "https:",
+						hostname: token.host,
+						port: "443",
+					},
+					userWorkerInspectorUrl: {
+						protocol: token.inspectorUrl.protocol,
+						hostname: token.inspectorUrl.hostname,
+						port: token.inspectorUrl.port.toString(),
+						pathname: token.inspectorUrl.pathname,
+					},
+					userWorkerInnerUrlOverrides: {
+						hostname: token.host,
+					},
+					headers: {
+						"cf-workers-preview-token": token.value,
+						...(accessToken
+							? { Cookie: `CF_Authorization=${accessToken}` }
+							: {}),
+					},
+					liveReload: config.dev.liveReload,
+					proxyLogsToController: true,
+					internalDurableObjects: [],
+					entrypointAddresses: {},
+				},
+			});
+		} catch (error) {
+			this.emitErrorEvent({
+				type: "error",
+				reason: "Error reloading remote server",
+				cause: castErrorCause(error),
+				source: "RemoteRuntimeController",
+				data: undefined,
+			});
+		}
+	}
+
 	// ******************
 	//   Event Handlers
 	// ******************
 
 	onBundleStart(_: BundleStartEvent) {
-		notImplemented(this.onBundleStart.name, this.constructor.name);
+		// Abort any previous operations when a new bundle is started
+		this.#abortController.abort();
+		this.#abortController = new AbortController();
 	}
-	onBundleComplete(_: BundleCompleteEvent) {
-		notImplemented(this.onBundleComplete.name, this.constructor.name);
+	onBundleComplete(ev: BundleCompleteEvent) {
+		const id = ++this.#currentBundleId;
+
+		if (!ev.config.dev?.remote) {
+			return;
+		}
+
+		this.emitReloadStartEvent({
+			type: "reloadStart",
+			config: ev.config,
+			bundle: ev.bundle,
+		});
+
+		void this.#mutex.runWith(() => this.#onBundleComplete(ev, id));
 	}
 	onPreviewTokenExpired(_: PreviewTokenExpiredEvent): void {
 		notImplemented(this.onPreviewTokenExpired.name, this.constructor.name);
 	}
 
 	async teardown() {
-		notImplemented(this.teardown.name, this.constructor.name);
+		// There's no way to teardown remote preview sessions, and so this is ignored in remote mode
```

Comment:
We should nullify the `#sessions` field so it can be garbage collected and not reused (although nothing should happen after a teardown but we might have some async callbacks which we don't bother cleaning up that might use it)

Reply:
Actually, we have an AbortController to abort

Reply:
83d914f61aed0379bb2a6ab26754f1314001df38